### DISCUSSION TITLE
Improving Performace

### DISCUSSION THREAD

**Discussion by Snolandia** [jkriege2/JKQtPlotter#148] [General]
Just wanted to start a discussion on this to keep track of what I'm finding/ trying to improve, and for suggestions/solutions.

This push started because while using the live-data sample graph, while the performance for 1 graph was ok enough for my uses,
while trying to add multiple graphs performance was significantly poorer than will meet my needs, as well as running into a weird rendering bug seemingly caused by the performance drop.

**Comment by Snolandia** [comment]
JKQTPEnhancedPainter, drawPolyLineFast

Im finding that QPainter drawPolyline has some signifigant performace issues. Just by moving to a loop and using QPainter::drawLine, Ive seen an average of 250% speed increase in drawing lines. Continuing to dig into this.

**Reply by Snolandia** [reply]
Continuing to use the same example. By only replacing drawPolylineFast, drawPolyline with the drawLines loops, and running a timing test over 1000 frames, ran several times for consistency, the results

No AA => Base: 22ms per frame.     Modified: 12ms per frame
AA       => Base: 83ms per frame,     Modified: 19ms per frame

while overall an obviously large improvement, the difference with AA turned on is absolutely massive, and was quite unexpected. QPainter::drawPolyline must have some major underlying issue with it.

**Reply by jkriege2** [reply]
I think one of the performance issues comes from the different ways to connect  the lines in a sequence:
https://doc.qt.io/qt-6/qt.html#PenJoinStyle-enum

I can imagine that this costs a lot of performance!

**Reply by jkriege2** [reply]
AntiAliasing is known to reduce performance ... therefore there are functions to enable/disable this explicitly in JKQTPlotter, see also:
https://jkriege2.github.io/JKQtPlotter/_j_k_q_t_plotter_speed_test.html

and:
https://jkriege2.github.io/JKQtPlotter/class_j_k_q_t_base_plotter.html#a4706bf608223fd2ffbf71afb5b9e5d4f
JKQTBasePlotter::setUseAntiAliasingForSystem(), setUseAntiAliasingForGraphs(),  setUseAntiAliasingForText()

**Reply by Snolandia** [reply]
Yes thats the example im doing testing on. For no modifications, the AntiAliasing being enabled is quite the performance hit, but when I replaced drawPolyline with a drawLines loop the performance hit from enabling AntiAliasing is quite small. No other changes were made for the testing. Testing was done with 2 examples running at the same time so they could be compared to each other while running. 

Im just trying to increase performance to keep up with some data feeds im trying to live stream on the graph.  

Next goal is setting up image caching so not everything has to be redrawn. 
For whatever reason, Qt Painter is really slow at drawing text, and ive been having issues with its performance in other parts of my program as well. The more I work with QtPainter, the more "interesting" quirks i keep discovering. Like trying to draw text rotated, which was a nightmare.

**Reply by jkriege2** [reply]
Yes, it might be a good idea to add functions that enable a mode where drawPolylineFast() is forced (if you'r OK with the reduced quality) ... that may be implemented alongside setUseAntiAliasingForGraphs() ...

**Reply by jkriege2** [reply]
image caching is something I wanted to implement at some point (do go ahead, if you want to!) ... 

text is - performance-wise - a weak point of JKQTPlotter: By using JKQTMathText it can draw using LaTeX markup, but also slows down the drawing .... but I decided FOR the plot quality, as performance was sufficient for me ...

**Reply by Snolandia** [reply]
> image caching is something I wanted to implement at some point (do go ahead, if you want to!) ...
> 
> text is - performance-wise - a weak point of JKQTPlotter: By using JKQTMathText it can draw using LaTeX markup, but also slows down the drawing .... but I decided FOR the plot quality, as performance was sufficient for me ...

Glad for the quality. For some tables I made, I spent way to long trying to get QPainter to draw text vertically correctly before I ended up just making it draw on a pixmap and rotated that. 

The image caching will probably take me a bit to set up as there is quite a lot of stuff intertwined to go through and finding the spare time to do it.

**Reply by Snolandia** [reply]
> Yes, it might be a good idea to add functions that enable a mode where drawPolylineFast() is forced (if you'r OK with the reduced quality) ... that may be implemented alongside setUseAntiAliasingForGraphs() ...

And as far as I could tell, there was no visual difference between drawPolylineFast, and drawLines. Though I didnt do any extensive testing.

**Reply by jkriege2** [reply]
One more thing about caching: there is some caching of text already ... there is already some caching for text drawing: not on the drawing itself, but for the size calculation. This is implemented in the get Mathtext Size functions in JKQTBasePlotter ...

This refines the need to parse the latex markup again and again and to recalculate the sizes of e.g. axis labels that do not change much between frames ...

**Reply by Snolandia** [reply]
Just giving an update. Caching has mostly been met with failure. Any sort of pixmap or caching of portions of the graph saw very little improvement over just drawing them outright, while causing significant hits to speed of portions calculated regulatory. The best I managed was bumping up to 45 fps from 30fps, running 2 of the example graphs at the same time. The only gains I managed were just not redrawing over portions that didnt have changes, rather than repainting the entire widget. For further performance gains using QPainter, I believe significant restructuring would be required, and QPainter just doesnt seem to have the capacity to do much better. I believe for major performance gains, restructuring to move to OpenGl would be the only realistic path forward, which is not a small undertaking unfortunately. 

In comparison, running 2 graphs with 2 plotted lines each with 500 visible datapoints, with 30,000 total datapoints in the loaded dataset through opengl, I get about 1000 frames per second.

**Reply by Snolandia** [reply]
Okay, update to my update.

Apparently Im missing something with QPainter or adjacent, because I was playing around with another library that uses QPainter only for plotting as well and with the same setup as in previous posts, it was running around 300 fps.  Ill have to dig around and see what the difference is if I get the chance.

**Comment by Snolandia** [comment]
Setting up SetX, SetY, and SetXY to have a No-ReDraw overload option, and using that when updating the multiple axis/datapoints then calling redraw at the end of everything significantly improved performance by nearly double for the live graph example, which setX then redrew the plot. framerate went from 30fps to about 60fps, based off a QTimer::singleshot set to 10ms after every frame update.

**Reply by jkriege2** [reply]
This is possible ... but it's not well documented:
Plotting is controlled via signals. There if you call JKQTPlotter::setPlotUpdateEnabled(false) before setting up the plot, redrawing is disabled and JJKQTPlotter::setPlotUpdateEnabled((true) reenables it. You may have to call redrawPlot() afterwards to update the display!

I just added a guard class JKQTPlotterUpdateGuard that automates this and improved the documentation about this topic!

**Reply by Snolandia** [reply]
Oh shoot. I totally missed that and going back and looking I see the function for it. Woooops.

**Comment by zhangxiaoGreen** [comment]
I want to generate a line Graph which has no more than 1000 line every line has 512 points. is this be supported
